crates/ty_ide/src/completion.rs

```diff
@@ -1,40 +1,784 @@
+        def foofoofoo(): ...
+    <CURSOR>
+
+def frob(): ...
+",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank5() {
+        let test = cursor_test(
+            "\
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+
+    <CURSOR>
+
+def frob(): ...
+",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn list_comprehension1() {
+        let test = cursor_test(
+            "\
+[<CURSOR> for bar in [1, 2, 3]]
+",
+        );
+
+        // It's not totally clear why `for` shows up in the
+        // symbol tables of the detected scopes here. My guess
+        // is that there's perhaps some sub-optimal behavior
+        // here because the list comprehension as written is not
+        // valid.
+        assert_snapshot!(test.completions(), @r"
+        for
+        bar
+        ");
+    }
+
+    #[test]
+    fn list_comprehension2() {
+        let test = cursor_test(
+            "\
+[f<CURSOR> for foo in [1, 2, 3]]
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foo
+        f
+        ");
+    }
+
+    #[test]
+    fn lambda_prefix1() {
+        let test = cursor_test(
+            "\
+(lambda foo: (1 + f<CURSOR> + 2))(2)
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foo
+        f
+        ");
+    }
+
+    #[test]
+    fn lambda_prefix2() {
+        let test = cursor_test(
+            "\
+(lambda foo: f<CURSOR> + 1)(2)
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foo
+        f
+        ");
+    }
+
+    #[test]
+    fn lambda_prefix3() {
+        let test = cursor_test(
+            "\
+(lambda foo: (f<CURSOR> + 1))(2)
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foo
+        f
+        ");
+    }
+
+    #[test]
+    fn lambda_prefix4() {
+        let test = cursor_test(
+            "\
+(lambda foo: 1 + f<CURSOR>)(2)
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foo
+        f
+        ");
+    }
+
+    #[test]
+    fn lambda_blank1() {
+        let test = cursor_test(
+            "\
+(lambda foo: 1 + <CURSOR> + 2)(2)
+",
+        );
+
+        assert_snapshot!(test.completions(), @"foo");
+    }
+
+    #[test]
+    fn lambda_blank2() {
+        let test = cursor_test(
+            "\
+(lambda foo: <CURSOR> + 1)(2)
+",
+        );
+
+        // FIXME: Should include `foo`.
+        //
+        // These fails for similar reasons as above: the body of the
+        // lambda doesn't include the position of <CURSOR> because
+        // <CURSOR> is inside leading or trailing whitespace. (Even
+        // when enclosed in parentheses.)
```

Comment:
Yeah, parentheses aren't part of the node's range unless it's relevant e.g., tuples.